crates/store/re_query/src/storage_engine.rs

```diff
@@ -0,0 +1,279 @@
+// Doing so would defeat the deadlock protection that the `StorageEngine` offers.
+// Exposing references to the actual `ChunkStore` and `QueryCache` if ofc fine.
+pub struct StorageEngineReadGuard<'a> {
+    store: parking_lot::RwLockReadGuard<'a, ChunkStore>,
+    cache: parking_lot::RwLockReadGuard<'a, QueryCache>,
+}
+
+impl Clone for StorageEngineReadGuard<'_> {
+    // Cloning the guard is safe, since the lock stays locked all along.
+    fn clone(&self) -> Self {
+        Self {
+            store: parking_lot::RwLock::read(RwLockReadGuard::rwlock(&self.store)),
+            cache: parking_lot::RwLock::read(RwLockReadGuard::rwlock(&self.cache)),
+        }
+    }
+}
+
+impl StorageEngineReadGuard<'_> {
+    #[inline]
+    pub fn store(&self) -> &ChunkStore {
+        &self.store
+    }
+
+    #[inline]
+    pub fn cache(&self) -> &QueryCache {
+        &self.cache
+    }
+}
+
+impl StorageEngineLike for StorageEngineReadGuard<'_> {
+    #[inline]
+    fn with_store<F: FnOnce(&ChunkStore) -> R, R>(&self, f: F) -> R {
+        f(self.store())
+    }
+
+    #[inline]
+    fn with_cache<F: FnOnce(&QueryCache) -> R, R>(&self, f: F) -> R {
+        f(self.cache())
+    }
+}
+
+// NOTE: None of these fields should ever be publicly exposed, either directly or through a method,
+// as it is always possible to go back to an actual `RwLock` via `ArcRwLockReadGuard::rwlock`.
+// Doing so would defeat the deadlock protection that the `StorageEngine` offers.
+// Exposing references to the actual `ChunkStore` and `QueryCache` if ofc fine.
+pub struct StorageEngineArcReadGuard {
+    store: parking_lot::ArcRwLockReadGuard<parking_lot::RawRwLock, ChunkStore>,
+    cache: parking_lot::ArcRwLockReadGuard<parking_lot::RawRwLock, QueryCache>,
+}
+
+impl StorageEngineArcReadGuard {
+    #[inline]
+    pub fn store(&self) -> &ChunkStore {
+        &self.store
+    }
+
+    #[inline]
+    pub fn cache(&self) -> &QueryCache {
+        &self.cache
+    }
+}
+
+impl StorageEngineLike for StorageEngineArcReadGuard {
+    #[inline]
+    fn with_store<F: FnOnce(&ChunkStore) -> R, R>(&self, f: F) -> R {
+        f(self.store())
+    }
+
+    #[inline]
+    fn with_cache<F: FnOnce(&QueryCache) -> R, R>(&self, f: F) -> R {
+        f(self.cache())
+    }
+}
+
+impl Clone for StorageEngineArcReadGuard {
+    // Cloning the guard is safe, since the lock stays locked all along.
+    fn clone(&self) -> Self {
+        Self {
+            store: parking_lot::RwLock::read_arc(ArcRwLockReadGuard::rwlock(&self.store)),
+            cache: parking_lot::RwLock::read_arc(ArcRwLockReadGuard::rwlock(&self.cache)),
+        }
+    }
+}
+
+// --- Write Guards ---
+
+// NOTE: None of these fields should ever be publicly exposed, either directly or through a method,
+// as it is always possible to go back to an actual `RwLock` via `RwLockWriteGuard::rwlock`.
+// Doing so would defeat the deadlock protection that the `StorageEngine` offers.
```

Comment:
perhaps it's worth documenting an example use case of doing the wrong thing without the protection that StorageEngine provides?